### PR TITLE
Introduce ov element type visitor

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Simplify evaluate implementation and code code maintenance by automatic code generation over switch case and macro
 - Applied visitor reduce openvinolib.so by around 27 KiB (Ubuntu 20.04, gcc 9.4, release build no LTO).

### Tickets:
 N/A

### CODE REVIEW

File: src/core/shape_inference/include/element_visitor.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+namespace ov {
+namespace element {
+
+/**
+ * @brief Defines supported elements for applying the element visitor action.
+ *
+ * @tparam List of supported ov::element types.
+ *
+ * The apply function check input element type, if is on the list apply function of visitor for specific type
```
Comment (t-jankowski) [inline src/core/shape_inference/include/element_visitor.hpp:20]: Does it regard only input elements? Or may it be used for just element type?

File: src/core/shape_inference/include/element_visitor.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+ * if not found apply default action.
```
Comment (t-jankowski) [inline src/core/shape_inference/include/element_visitor.hpp:21]:
```suggestion
 * The apply function checks -input- element type whether it is on the list ?of? apply function of visitor for specific type
 * if not found default action is applied.
```

File: src/core/shape_inference/include/element_visitor.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+struct NoAction<void> {
+    using result_type = void;
+
+    static void visit() {}
+};
+
+/**
+ * @brief Helper visitor which throws ov::Exception for not supported element type.
+ *
+ * @tparam R Type of return type (used to be compatible with others call operator in Visitor).
```
Comment (t-jankowski) [inline src/core/shape_inference/include/element_visitor.hpp:102]:
```suggestion
 * @tparam R Type of return type (used to be compatible with other call operator in Visitor).
```

File: src/core/shape_inference/include/element_visitor.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    template <class Visitor, class... Args>
+    static auto apply(Type_t et, Args&&... args) -> typename Visitor::result_type {
+        return (et == ET) ? Visitor::template visit<ET>(std::forward<Args>(args)...)
+                          : IfTypeOf<Others...>::template apply<Visitor>(et, std::forward<Args>(args)...);
+    }
+};
+
+/**
+ * @brief Helper visitor which define no action for not supported type.
```
Comment (t-jankowski) [inline src/core/shape_inference/include/element_visitor.hpp:73]:
```suggestion
 * @brief Helper visitor which defines no action for not supported type.
```

File: src/core/shape_inference/include/element_visitor.hpp (hunk 2 above)
Comment (praasz) [reply src/core/shape_inference/include/element_visitor.hpp]: Comment updated.

File: src/core/shape_inference/include/element_visitor.hpp (hunk 1 above)
Comment (praasz) [reply src/core/shape_inference/include/element_visitor.hpp]: Comment updated.

File: src/core/shape_inference/include/element_visitor.hpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+/**
+ * @brief Helper visitor which defines no action for not supported type.
+ *
+ * @tparam R     Type of return value.
+ * @tparam value Default value returned.
+ */
+template <class R, R... value>
+struct NoAction {
+    static_assert(sizeof...(value) < 2, "There should no more then one result value.");
```
Comment (t-jankowski) [inline src/core/shape_inference/include/element_visitor.hpp:84]:
```suggestion
    static_assert(sizeof...(value) < 2, "There should be no more than one result value.");
```

File: src/core/shape_inference/include/element_visitor.hpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    using result_type = R;
+
+    static constexpr result_type visit() {
+        return {value...};
+    }
+};
+
+/**
+ * @brief Helper visitor which define no action for not supported type if result is void type.
```
Comment (t-jankowski) [inline src/core/shape_inference/include/element_visitor.hpp:94]:
```suggestion
 * @brief Helper visitor which defines no action for not supported type if result is void type.
```

File: src/core/src/op/round.cpp [hunk 7]
```diff
@@ -15,50 +16,48 @@ using namespace std;
...
-// function used by COPY_TENSOR
-template <element::Type_t ET>
-inline bool copy_tensor(const HostTensorPtr& arg0, const HostTensorPtr& out, const size_t count) {
-    runtime::reference::copy(arg0->get_data_ptr<ET>(), out->get_data_ptr<ET>(), count);
-    return true;
-}
+class Evaluate : public ov::element::NoAction<bool> {
+    template <element::Type_t ET>
+    static constexpr bool is_floating() {
+        return (ET == element::f16) || (ET == element::f32) || (ET == element::bf16);
```
Comment (t-jankowski) [inline src/core/src/op/round.cpp:23]: What about f64? At least a comment why not.

File: src/plugins/intel_cpu/tests/unit/shape_inference_test/element_visitor_test.cpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2023 Intel Corporation
```
Comment (t-jankowski) [inline src/plugins/intel_cpu/tests/unit/shape_inference_test/element_visitor_test.cpp:1]: Is plugin right place for these tests?

Comment (praasz) [reply src/plugins/intel_cpu/tests/unit/shape_inference_test/element_visitor_test.cpp]: Moved to core

**Comment by mlukasze** [comment]
@ilyachur & @mitruska could you review it, please?

File: src/core/src/op/round.cpp [hunk 9]
```diff
@@ -15,50 +16,48 @@ using namespace std;
...
+    }
 
+public:
+    using ov::element::NoAction<bool>::visit;
+    template <element::Type_t ET, typename std::enable_if<!is_floating<ET>()>::type* = nullptr>
+    static result_type visit(const HostTensorPtr& arg0,
+                             const HostTensorPtr& out,
+                             const size_t count,
+                             const op::v5::Round::RoundMode) {
+        memcpy(out->get_data_ptr(), arg0->get_data_ptr(), out->get_size_in_bytes());
```
Comment (ilyachur) [inline src/core/src/op/round.cpp:33]: From my point of view will be better to have this differentiation inside `ngraph::runtime::reference::round()` function. Inside this function we need to select the optimal implementation.

**Review by ilyachur — APPROVED** [review]
The idea LGTM

File: src/core/src/op/round.cpp (hunk 9 above)
Comment (praasz) [reply src/core/src/op/round.cpp:33]: Agree, created ticket to refactor reference implementation and apply this change.

**Pull request merged**